mflowgen/bin/setup-buildkite.sh

```diff
@@ -245,6 +244,30 @@ else
 fi
 
 
+##############################################################################
+# LOCK so that no two script instances can run at the same time.
+# In particular, do not want e.g. competing 'git clone' or
+# 'pip install' ops trying to access the same directory etc.
+
+echo "--- LOCK"
+LOCK=/tmp/setup-buildkite.lock
+exec 9>> $LOCK
+date; echo "I am process $$ and I want lock '$LOCK'"
+if ! flock -n 9; then
+    echo "Waiting for process `cat $LOCK` to release the lock..."
+    if ! flock -w 600 9; then
+        echo "ERROR waited ten minutes and could not get lock '$LOCK'"
+        echo "apparently held by process `cat $LOCK`"
+        return 13 || exit 13
+    fi
+fi
+date; echo -n "Lock acquired! Prev owner was "; cat $LOCK
+echo $$ > $LOCK; # Record who has the lock (i.e. me)
+
+# Failsafe: Release lock on exit (if not before).  Note, because this
+# script is sourced, this trap won't kick in until calling process dies.
+trap "flock -u 9" EXIT
+
```

Comment:
I don't really understand why this is needed... could you explain more?

Reply:
Basically why are parallel jobs trying to access the same directory? Shouldn't they just be completely independent, separate parallel jobs?

Reply:
> Basically why are parallel jobs trying to modify the same directory? Shouldn't they just be completely independent, separate parallel jobs?

My setup script was designed to do one job at a time, and was doing dumb things like the following:
```
test -e adks || ln -s /sim/buildkite-agent/adks
cd adks/tsmc16-adk; git pull
```
Without the `flock`, when two tests run the setup script at the same time, and both try to do a pull from the same repo, one of them gets a collision error and fails. There are several such places where a "git pull" or a "git checkout" or a "pip install" etc. had the same potential problems...for awhile I was trying to track each one down and do the right thing so there would be no collision...in the end it was simpler to just implement the `flock` after which everything worked as expected...